quaint/src/connector/postgres/native/websocket.rs

```diff
@@ -95,3 +109,120 @@ impl From<TungsteniteError> for error::Error {
         builder.build()
     }
 }
+
+#[pin_project]
+struct WsTunnel {
+    #[pin]
+    inner: StreamReader<WsBytesStream, Bytes>,
+    write_state: WriteState,
+}
+
+enum WriteState {
+    Free,
+    Writing(usize),
+}
+
+#[pin_project]
+struct WsBytesStream(#[pin] WebSocketStream<MaybeTlsStream<TcpStream>>);
+
+impl WsTunnel {
+    fn new(stream: WebSocketStream<MaybeTlsStream<TcpStream>>) -> Self {
+        WsTunnel {
+            inner: StreamReader::new(WsBytesStream(stream)),
+            write_state: WriteState::Free,
+        }
+    }
+}
+
+impl WsBytesStream {
+    fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut WebSocketStream<MaybeTlsStream<TcpStream>>> {
+        self.project().0
+    }
+}
+
+impl AsyncRead for WsTunnel {
+    fn poll_read(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &mut ReadBuf<'_>) -> Poll<std::io::Result<()>> {
+        self.project().inner.poll_read(cx, buf)
+    }
+}
+
+impl AsyncBufRead for WsTunnel {
+    fn poll_fill_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<&[u8]>> {
+        self.project().inner.poll_fill_buf(cx)
+    }
+
+    fn consume(self: Pin<&mut Self>, amt: usize) {
+        self.project().inner.consume(amt)
+    }
+}
+
+impl AsyncWrite for WsTunnel {
+    fn poll_write(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<std::io::Result<usize>> {
+        let this = self.get_mut();
+        let sink = &mut this.inner.get_mut().0;
+        let to_io_err = |err| IoError::new(IoErrorKind::Other, err);
+
+        match this.write_state {
+            WriteState::Free => {
+                ready!(sink.poll_ready_unpin(cx)).map_err(to_io_err)?;
+                sink.start_send_unpin(Message::Binary(Bytes::copy_from_slice(buf)))
+                    .map_err(to_io_err)?;
+                this.write_state = WriteState::Writing(buf.len());
+                cx.waker().wake_by_ref();
+                Poll::Pending
```

Comment:
Technically we could skip an extra roundtrip back to the executor and start polling flush right away, but the code looks a bit nicer without this micro-optimization so 🤷